extensions/amp-access/0.1/access-expr.js

```diff
@@ -44,3 +45,51 @@ export function evaluateAccessExpr(expr, data) {
     parser.yy = null;
   }
 }
+
+/**
+ * AmpAccessEvaluator evaluates amp-access expressions.
+ * It uses a cache to speed up repeated evals for the same expression.
+ */
+export class AmpAccessEvaluator {
+  /** */
+  constructor() {
+    /** @type {Object<string, boolean>} */
+    this.cache = null;
+
+    /** @type {JsonObject} */
+    this.lastData = null;
+  }
+
+  /**
+   * Evaluate access expressions, but turn to a cache first.
+   * Cache is invalidated when given new data.
+   *
+   * @param {string} expr
+   * @param {!JsonObject} data
+   * @return {boolean}
+   */
+  evaluate(expr, data) {
+    if (this.lastData !== data) {
+      this.lastData = data;
+      this.cache = map();
+    }
+
+    if (!hasOwn(this.cache, expr)) {
```

Comment:
You don't need `hasOwn` check with a pure map object.

Reply:
what should I used instead? `in`?

Reply:
Just check if there's a property? Are they evals falsey?

Reply:
I had done that initially, but seeing as the return value is of type boolean, they are often falsey :). This led to many cache misses